src/game/client/components/binds.cpp

```diff
@@ -25,82 +25,119 @@ const char CBinds::s_aaDefaultBindValues[][32] = {
+	str_copy(m_aaaKeyBindings[KeyID][Modifier], pStr, sizeof(m_aaaKeyBindings[KeyID][Modifier]));
 	char aBuf[256];
-	if(!m_aaKeyBindings[KeyID][0])
-		str_format(aBuf, sizeof(aBuf), "unbound %s (%d)", Input()->KeyName(KeyID), KeyID);
+	if(!m_aaaKeyBindings[KeyID][Modifier][0])
+		str_format(aBuf, sizeof(aBuf), "unbound %s%s (%d)", GetModifierName(Modifier),Input()->KeyName(KeyID), KeyID);
 	else
-		str_format(aBuf, sizeof(aBuf), "bound %s (%d) = %s", Input()->KeyName(KeyID), KeyID, m_aaKeyBindings[KeyID]);
+		str_format(aBuf, sizeof(aBuf), "bound %s%s (%d) = %s", GetModifierName(Modifier),Input()->KeyName(KeyID), KeyID, m_aaaKeyBindings[KeyID][Modifier]);
 	Console()->Print(IConsole::OUTPUT_LEVEL_STANDARD, "binds", aBuf);
 }
 
+bool CBinds::CBindsSpecial::OnInput(IInput::CEvent Event)
+{
+	int Mask = 0;
+	// since we only handle one modifier, when doing ctrl+q and shift+q, execute both
+	Mask |= Input()->KeyIsPressed(KEY_LSHIFT) << MODIFIER_SHIFT;
+	Mask |= Input()->KeyIsPressed(KEY_RSHIFT) << MODIFIER_SHIFT;
+	Mask |= Input()->KeyIsPressed(KEY_LCTRL) << MODIFIER_CTRL;
+	Mask |= Input()->KeyIsPressed(KEY_RCTRL) << MODIFIER_CTRL;
+	Mask |= Input()->KeyIsPressed(KEY_LALT) << MODIFIER_ALT;
+	if(Mask == 0) 
+		Mask = 1; // if no modifier, flag with MODIFIER_NONE
+	bool rtn = false;
+
+	// don't handle invalid events and keys that aren't set to anything
+	if(!((Event.m_Key >= KEY_F1 && Event.m_Key <= KEY_F12) || (Event.m_Key >= KEY_F13 && Event.m_Key <= KEY_F24)))
+		return false;
+
+	for(int m = 0; m < MODIFIER_COUNT; m++)
+	{
+		if((Mask&(1 << m)) && m_pBinds->m_aaaKeyBindings[Event.m_Key][m][0] != 0)
+		{
+			int Stroke = 0;
+			if(Event.m_Flags&IInput::FLAG_PRESS)
+				Stroke = 1;
+
+			m_pBinds->GetConsole()->ExecuteLineStroked(Stroke, m_pBinds->m_aaaKeyBindings[Event.m_Key][m]);
+			rtn = true;
+		}
+	}
+	return rtn;
+}
 
 bool CBinds::OnInput(IInput::CEvent e)
 {
-	// don't handle invalid events and keys that arn't set to anything
-	if(e.m_Key <= 0 || e.m_Key >= KEY_LAST || m_aaKeyBindings[e.m_Key][0] == 0)
+	int Mask = 0;
+	// since we only handle one modifier, when doing ctrl+q and shift+q, execute both
+	Mask |= Input()->KeyIsPressed(KEY_LSHIFT) << MODIFIER_SHIFT;
+	Mask |= Input()->KeyIsPressed(KEY_RSHIFT) << MODIFIER_SHIFT;
+	Mask |= Input()->KeyIsPressed(KEY_LCTRL) << MODIFIER_CTRL;
+	Mask |= Input()->KeyIsPressed(KEY_RCTRL) << MODIFIER_CTRL;
+	Mask |= Input()->KeyIsPressed(KEY_LALT) << MODIFIER_ALT;
+	if(Mask == 0) 
+		Mask = 1; // if no modifier, flag with MODIFIER_NONE
+
+	// don't handle invalid events and keys that aren't set to anything
+	if(e.m_Key <= 0 || e.m_Key >= KEY_LAST)
 		return false;
 
-	if(e.m_Flags&IInput::FLAG_PRESS)
-		Console()->ExecuteLineStroked(1, m_aaKeyBindings[e.m_Key]);
-	if(e.m_Flags&IInput::FLAG_RELEASE)
-		Console()->ExecuteLineStroked(0, m_aaKeyBindings[e.m_Key]);
-	return true;
+	bool rtn = false;
+	for(int m = 0; m < MODIFIER_COUNT; m++)
+	{
+		if((Mask&(1 << m)) && m_aaaKeyBindings[e.m_Key][m][0] == 0)
```

Comment:
This is wrong and should be
`if((Mask&(1 << m)) && m_pBinds->m_aaaKeyBindings[Event.m_Key][m][0] != 0)
{
...
}`